ctl/shared_ptr.cc

```diff
@@ -0,0 +1,97 @@
+// -*- mode:c++; indent-tabs-mode:nil; c-basic-offset:4; coding:utf-8 -*-
+// vi: set et ft=cpp ts=4 sts=4 sw=4 fenc=utf-8 :vi
+//
+// Copyright 2024 Justine Alexandra Roberts Tunney
+//
+// Permission to use, copy, modify, and/or distribute this software for
+// any purpose with or without fee is hereby granted, provided that the
+// above copyright notice and this permission notice appear in all copies.
+//
+// THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL
+// WARRANTIES WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED
+// WARRANTIES OF MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE
+// AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL
+// DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR
+// PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER
+// TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
+// PERFORMANCE OF THIS SOFTWARE.
+
+#include "shared_ptr.h"
+
+#include "libc/intrin/atomic.h"
+
+namespace {
+
+inline void
+incref(_Atomic(size_t)* r)
+{
+    size_t r2 = atomic_fetch_add_explicit(r, 1, memory_order_relaxed);
+    if (r2 > ((size_t)-1) >> 1)
+        __builtin_trap();
+}
+
+inline int
+decref(_Atomic(size_t)* r)
+{
+    if (!atomic_fetch_sub_explicit(r, 1, memory_order_release)) {
+        atomic_thread_fence(memory_order_acquire);
```

Comment:
I'm not sure how I feel about `atomic_thread_fence()`. I just tried using a similar trick to this for my syncthreads() function on an M2 processor and it made things slower. For example:

```c
void
syncthreads(void)
{
    static atomic_uint count;
    static atomic_uint phaser;
    int phase = atomic_load_explicit(&phaser, memory_order_relaxed);
    if (atomic_fetch_add_explicit(&count, 1, memory_order_acq_rel) + 1 == nth) {
        atomic_store_explicit(&count, 0, memory_order_relaxed);
        atomic_store_explicit(&phaser, phase + 1, memory_order_release);
    } else {
        int backoff = 0;
        while (atomic_load_explicit(&phaser, memory_order_acquire) == phase)
            backoff = delay(backoff);
    }
}
```

Became:

```c
void
syncthreads(void)
{
    static atomic_uint count;
    static atomic_uint phaser;
    int phase = atomic_load_explicit(&phaser, memory_order_relaxed);
    if (atomic_fetch_add_explicit(&count, 1, memory_order_release) + 1 == nth) {
        atomic_thread_fence(memory_order_acquire);
        atomic_store_explicit(&count, 0, memory_order_relaxed);
        atomic_store_explicit(&phaser, phase + 1, memory_order_release);
    } else {
        int backoff = 0;
        while (atomic_load_explicit(&phaser, memory_order_acquire) == phase)
            backoff = delay(backoff);
    }
}
```

One nasty thing about `atomic_thread_fence()` is it isn't supported by TSAN at all. So it's harder to prove code is correct.

How certain are you that this decref() implementation is optimal? Could you whip up a torture test + benchmark that demonstrates its merit in comparison to alternatives? I would have assumed that that doing an acquire load beforehand to check for zero would be faster, since it'd let you avoid the costly `XADD` instruction in many cases.

Reply:
Not certain at all, would be interesting to see it benchmarked at some point; I just cribbed off of Rust, and Rust is pretty good on these fronts but not perfect.

If acq_rel is opitmal then that's a point in libc++'s favor...

Reply:
Ignore my earlier comment, I've discovered how to correctly use atomic thread fence. Consider the following barrier, where `nth` is the number of threads and `ith` is the thread id:

```c
void
syncthreads(int ith)
{
    static atomic_uint count;
    static atomic_uint phaser;
    int phase = atomic_load_explicit(&phaser, memory_order_relaxed);
    if (atomic_fetch_add_explicit(&count, 1, memory_order_acq_rel) + 1 == nth) {
        atomic_store_explicit(&count, 0, memory_order_relaxed);
        atomic_store_explicit(&phaser, phase + 1, memory_order_release);
    } else {
        int backoff = 0;
        while (atomic_load_explicit(&phaser, memory_order_acquire) == phase)
            backoff = delay(backoff);
    }
}
```

Spinning on separate cache lines and using thread fence brings its latency down from 7.3 us to 6.2 us on my M2:

```c
struct ggml_phaser
{
    alignas(64) atomic_uint i;
};

void
syncthreads(int ith)
{
    static atomic_uint count;
    static struct ggml_phaser ph[nth];
    int phase = atomic_load_explicit(&ph[ith].i, memory_order_relaxed);
    if (atomic_fetch_add_explicit(&count, 1, memory_order_acq_rel) + 1 == nth) {
        atomic_store_explicit(&count, 0, memory_order_relaxed);
        for (int i = 0; i < nth; ++i)
            atomic_store_explicit(&ph[i].i, phase + 1, memory_order_relaxed);
        atomic_thread_fence(memory_order_release);
    } else {
        int backoff = 0;
        while (atomic_load_explicit(&ph[ith].i, memory_order_relaxed) == phase)
            backoff = delay(backoff);
        atomic_thread_fence(memory_order_acquire);
    }
}
```